ibis/backends/decompile/compiler.py

```diff
@@ -0,0 +1,145 @@
+import functools
+import io
+import itertools
+
+import black
+
+import ibis.common.graph as graph
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.schema as sch
+
+
+@functools.singledispatch
+def translate(op, *args, **kwargs):
+    raise NotImplementedError(op)
+
+
+@translate.register(dt.DataType)
+def datatype(dtype, *args, **kwargs):
+    return str(dtype)
+
+
+@translate.register(sch.Schema)
+def schema(schema, *, names, types):
+    return repr(list(zip(names, map(str, types))))
+
+
+@translate.register(ops.UnboundTable)
+def table(op, schema, name):
+    return f"ibis.table(name={name!r}, schema={schema})"
+
+
+@translate.register(ops.Selection)
+def selection(op, table, selections, predicates, sort_keys):
+    out = f"{table}"
+    if selections:
+        out = f"{out}.select({selections})"
+    if predicates:
+        out = f"{out}.filter({predicates})"
+    if sort_keys:
+        out = f"{out}.sort_by({sort_keys})"
+    return out
+
+
+@translate.register(ops.Limit)
+def limit(op, table, n, offset):
+    if offset:
+        return f"{table}.limit({n}, {offset})"
+    else:
+        return f"{table}.limit({n})"
+
+
+@translate.register(ops.TableColumn)
+def table_column(op, table, name):
+    return f"{table}.{name}"
+
+
+@translate.register(ops.SortKey)
+def sort_key(op, expr, ascending):
+    if ascending:
+        return f"ibis.asc({expr})"
+    else:
+        return f"ibis.desc({expr})"
+
+
+@translate.register(ops.Reduction)
+def reduction(op, arg, where, **kwargs):
+    name = op.__class__.__name__.lower()
+    return f"{arg}.{name}()"
+
+
+@translate.register(ops.Alias)
+def alias(op, arg, name):
+    return f"{arg}.name({name!r})"
+
+
+@translate.register(ops.Literal)
+def literal(op, value, dtype):
+    return repr(value)
+
+
+@translate.register(ops.NodeList)
+def node_list(op, *values):
+    return ", ".join(values)
+
+
+@translate.register(ops.Equals)
+def equals(op, left, right):
+    return f"{left} == {right}"
+
+
+@translate.register(ops.Divide)
```

Comment:
Eventually the arithmetic operations will have to be parenthesized. I know you haven't implemented `Add` or `Subtract` yet, but you won't be able to just render the operands recursively.

Consider this case:

```
In [11]: (ibis.literal(2) - ibis.literal(3)) + 4
Out[11]: Add(Subtract(2, 3), 4): 2 - 3 + 4

In [12]: ibis.literal(2) - (ibis.literal(3) + 4)
Out[12]: Subtract(2, Add(3, 4)): 2 - 3 + 4
```

We're technically rendering this incorrectly in the repr right now, and you'll have to deal with the parenthesization problem here too.

Reply:
I'm handling parenthesization from parent infix operators now, seems to work fine.